tests/api_workflow/test_api_workflow_datasets.py

```diff
@@ -277,6 +277,43 @@ def test_get_datasets__shared_None(mocker: MockerFixture) -> None:
     assert mocked_pagination.call_count == 3
 
 
+def test_get_datasets_by_name__paginated(mocker: MockerFixture) -> None:
+    datasets = _get_datasets(3)
+    # Returns the same set of datasets twice. API client should remove duplicates
+    mocker.patch.object(
+        api_workflow_datasets.utils,
+        "paginate_endpoint",
+        side_effect=[
+            # shared=False, simulate two pages.
+            iter([datasets[0], datasets[1]]),
+            # shared=True, simulate one page.
+            iter([datasets[2]]),
+            iter([]),
+            # shared=None
+            iter([datasets[0], datasets[1]]),
+            iter([datasets[2]]),
+            iter([]),
+        ],
+    )
+    mocker.patch.object(ApiWorkflowClient, "__init__", return_value=None)
+    mock_datasets_api = mocker.MagicMock()
+    client = ApiWorkflowClient()
+    client._datasets_api = mock_datasets_api
+
+    datasets_not_shared = client.get_datasets_by_name(
+        shared=False, dataset_name="mocked_dataset_0"
+    )
```

Comment:
In general, this is quite hard to understand and I am not sure if the tests are both correct and cover all cases.
Some points:

- where does the magic `mocked_dataset_0` come from? Can I replace it with any other string? If not, then this should be clear by the code.
- If I set `shared=None` or `shared=True` here, will the test fail? I don't see how it relates to the mocking. 
- Instead of mocking once for all 3 cases of shared, why not mock before each case? That would make the relationship clear in code, not only in comments.
- Is the goal to test the `get_datasets_by_name` method or the `paginated` function or both?

Reply:
As far as I see, your changes addressed point 3, but not the other ones.

Reply:
Such a test fulfils the requirements:

```
def test_get_datasets_by_name__not_shared__paginated(mocker: MockerFixture) -> None:
    datasets = _get_datasets(3)
    mocked_paginate_endpoint = mocker.patch.object(
        api_workflow_datasets.utils,
        "paginate_endpoint",
        return_value=iter([datasets[0]]),
    )
    mocker.patch.object(ApiWorkflowClient, "__init__", return_value=None)
    mock_datasets_api = mocker.MagicMock()
    client = ApiWorkflowClient()
    client._datasets_api = mock_datasets_api

    # Note: because the `dataset_name` filtering is mocked away in this test,
    # the `dataset_name` passed as argument and in the returned dataset are independent.
    datasets_returned = client.get_datasets_by_name(
        shared=False, 
        dataset_name="some_random_dataset_name"
    )
    assert datasets_returned == [datasets[0]]

    mocked_paginate_endpoint.assert_called_once_with(
        mock_datasets_api.get_datasets_query_by_name,
        dataset_name="some_random_dataset_name",
        exact=True,
        shared=False
    )
```